src/extensions/default/AutoUpdate/main.js

```diff
@@ -0,0 +1,608 @@
+                    description: Strings.GO_TO_SITE
+                });
+            }
+        } else if (downloadCompleted && !updateInitiatedInPrevSession) {
+            // We get here if the download was complete and user selected UpdateLater
+            if (brackets.platform === "mac") {
+                filesToCache = ['.dmg', '.json'];
+            } else if (brackets.platform === "win") {
+                filesToCache = ['.msi', '.json'];
+            }
+        }
+
+        postMessageToNode(MessageIds.PERFORM_CLEANUP, filesToCache);
+    }
+
+    /**
+     * Initializes the state for AutoUpdate process
+     * @returns {$.Deferred} - a jquery promise,
+     *                       that is resolved with success or failure
+     *                       of state initialization
+     */
+    function initializeState() {
+        var result = $.Deferred();
+
+        FileSystem.resolve(updateDir, function (err) {
+            if (!err) {
+                result.resolve();
+            } else {
+                var directory = FileSystem.getDirectoryForPath(updateDir);
+                directory.create(function (error) {
+                    if (error) {
+                        console.error('AutoUpdate : Error in creating update directory in Appdata');
+                        result.reject();
+                    } else {
+                        result.resolve();
+                    }
+                });
+            }
+        });
+
+        return result.promise();
+    }
+
+    /**
+     * Enables/disables the state of "Check For Updates" menu entry under Help Menu
+     */
+    function enableCheckForUpdateEntry(enable) {
+        var cfuCommand = CommandManager.get(Commands.HELP_CHECK_FOR_UPDATE);
+        cfuCommand.setEnabled(enable);
+    }
+
+    /**
+     * Checks if it is the first iteration of download
+     * @returns {boolean} - true if first iteration, false if is a retrial of download
+     */
+    function isFirstIterationDownload() {
+        return (downloadAttemptsRemaining === MAX_DOWNLOAD_ATTEMPTS);
+    }
+
+    /**
+     * Resets the update state in updatehelper.json in case of failure,
+     * and logs an error with the message
+     * @param {string} message - the message to be logged onto console
+     */
+    function resetStateInFailure(message) {
+        updateJsonHandler.reset();
+        UpdateInfoBar.showUpdateBar({
+            type: "error",
+            title: Strings.UPDATE_FAILED,
+            description: ""
+        });
+        console.error(message);
+    }
+
+    /**
+     * Sets the update state in updateHelper.json in Appdata
+     * @param {string} key   - key to be set
+     * @param {string} value  - value to be set
+     * @param {function} callback - the callback function
+     *                            to be called in case of
+     *                            successful setting of update state
+     */
+    function setUpdateStateInJSON(key, value, callback) {
```

Comment:
This is indeed a promise based function. The same promise handling code snippet was getting used at multiple places, and so, in order to avoid the repetition of code, I have templatized the code snippet.